src/compiler/tsbuildPublic.ts

```diff
@@ -861,6 +871,11 @@ namespace ts {
                 getConfigFileParsingDiagnostics(config),
                 config.projectReferences
             );
+            state.lastCachedPackageJsonLookups.set(projectPath, state.moduleResolutionCache && map(
+                state.moduleResolutionCache.getPackageJsonInfoCache().entries(),
+                ([path, data]) => ([state.host.realpath ? state.host.realpath(path) as Path : path, data] as const)

```

Comment:
The `realpath` of a `Path` could be not-a-`Path`, no?

Reply:
A `Path` is a canonically cased absolute path - realpath should definitely return an absolute path, and it should return the FS case of the path (or the input case, depending on platform and case sensitivity), which is arguably even more canonical. It doesn't feel right to take the output of realpath (the most canonical version of a path we can get) and potentially mangle it back into our idea of a `Path`.

Reply:
It potentially matters, though, because this value eventually becomes a key of `allWatchedPackageJsonFiles`, which are used to prevent duplication of watches. On a case-insensitive file system, the realpath could come back with a mixed-case filename, and then (maybe???) another resolution to that same realpath could turn up, correctly `Path`ified, in the module resolution cache, ultimately causing you to create one watcher for the all-lowercased `Path` and one for the mixed-case original filename. Ultimately, if filenames are going to be keys in maps, they need to be `Path`ified.

Reply:
Every entry in that table gets `realpath`'d (if it exists) though, so there won't be duplicate entries, at least. And since we wipe away and replace the whole table (elementwise, via mutateMap), we don't even run the risk of accidentally leaving behind old watches if there's a case mismatch (we'd just remove the old case, then add the new one, which is potentially inefficient, but not the worst). And again, at least on windows, node realpath returns input case, so it shouldn't change anything (unless it remaps a symlink, in which case the new part will be FS case, I think). (If you always wanted the FS case you actually have to use the "native" realpath variant, iirc, which we don't use because it caused issues if I'm remembering correctly).

Reply:
> unless it remaps a symlink, in which case the new part will be FS case, I think

This is what I was talking about. And because _otherwise_ `realpath` returns input case, it seems like such a remapping _could_ result in duplicate watches. At any rate, if you’re completely sure that it doesn’t matter what case the keys are (though it seems like it does), then type them as a `string`, because the type assertion is falsifiable.

Reply:
Given that it looks like we're trying to use `realpath.native` (which always returns FS case, iirc) [here](https://github.com/microsoft/TypeScript/pull/44966), I've wrapped this with `toPath` calls like you ask, so we won't have to come back here and addjust this again later.